File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
    size_t left = 0;
    size_t right = length;
 
    while (left < right) {
        size_t mid = left + (right - left) / 2;

        if (arr[mid] == value)
            return &arr[mid];

        if (arr[mid] < value)
            left = mid + 1;
        else
            right = mid;
    }

    return NULL;
}

File: solutions/ruby/binary-search/1/binary_search.rb
class BinarySearch
  def initialize(values)
    @values = values
  end

  def search_for(n)
    left = 0
    right = @values.length

    while left < right
      mid = left + (right - left) / 2

      return mid if @values[mid] == n

      if @values[mid] < n 
        left = mid + 1
      else
        right = mid
      end
    end
    nil
  end
end